### ISSUE TITLE
TensorRT integration doesn't optimize conv2d_transpose

### ISSUE THREAD

**Issue by qinyao-he** [tensorflow/tensorflow#20157]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**:
Yes.
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:
Ubuntu 16.04
- **TensorFlow installed from (source or binary)**:
Source. Build toward TensorRT 3.0.4.
- **TensorFlow version (use command below)**:
1.7
- **Python version**: 
2.7.12
- **Bazel version (if compiling from source)**:
0.11.1
- **GCC/Compiler version (if compiling from source)**:
5.4.0
- **CUDA/cuDNN version**:
CUDA 9.0, cuDNN 7.1.4
- **GPU model and memory**:
NVIDIA 1080 Ti
- **Exact command to reproduce**:
First train a example model use [mnist_conv_deconv.py](https://gist.github.com/qinyao-he/573254f8e225a09114b7408cc1c984aa)
Then use tensorflow built-in tools to freeze the graph:
```
python -m tensorflow.python.tools.freeze_graph --input_graph log/graph.pbtxt --input_checkpoint log/model.ckpt-20000 --output_node_names L2Loss --output_graph log/freeze_graph.pb
```
Finally use [tensorrt.py](https://gist.github.com/qinyao-he/28ddedb7f561bb3cb4ba880833f14a89) to optimize the graph use TensorRT engine.

### Describe the problem
The original graph contains 4 convolution and 4 deconvolution layers.
<img width="179" alt="snipaste_2018-06-20_12-39-15" src="https://user-images.githubusercontent.com/6523975/41680987-1e912d1e-7488-11e8-8a25-6f8e112cc2e6.png">

After optimizing, convolution layers was converted into trt_op (successfully optimized by TensorRT), while all deconvolution layers (conv2d_transpose) remains unchanged.
<img width="244" alt="snipaste_2018-06-20_12-39-46" src="https://user-images.githubusercontent.com/6523975/41681041-4bba6a08-7488-11e8-970c-2773fedea98a.png">

I think TensorRT support deconvolution as shown in their official guide [TensorRT Developer Guide](https://docs.nvidia.com/deeplearning/sdk/tensorrt-developer-guide/index.html)

> TensorFlow
> The following list describes the operations that are supported in a TensorFlow framework.
> Placeholder
> Const
> Add, Sub, Mul, Div, Minimum and Maximum
> BiasAdd
> Negative, Abs, Sqrt, Rsqrt, Pow, Exp and Log
> FusedBatchNorm
> ReLU, TanH, Sigmoid
> SoftMax
> Mean
> ConcatV2
> Reshape
> Transpose
> Conv2D
> DepthwiseConv2dNative
> ConvTranspose2D
> MaxPool
> AvgPool
> Pad is supported if followed by one of these TensorFlow layers: Conv2D, DepthwiseConv2dNative, MaxPool, and AvgPool

So is there any problem in tensorflow integrations?

**Comment by qinyao-he** [comment]
@samikama

**Comment by samikama** [comment]
@qinyao-he There could be many reasons but it is not possible to figure out what is wrong from the pictures since only parent scopes are visible in the plots but not the nodes themselves. Would it be possible to share the network? At least some logs? Also for the time being, TFTRT requires shapes to be statically inferable at the conversion time. Are your non-batch input ranks fixed or are the free?

**Comment by qinyao-he** [comment]
https://gist.github.com/qinyao-he/573254f8e225a09114b7408cc1c984aa
https://gist.github.com/qinyao-he/28ddedb7f561bb3cb4ba880833f14a89
@samikama These are the two code snippet to reproduce as I stated in the issue description.

**Comment by qinyao-he** [comment]
@zheng-xq

**Comment by xiaocai00** [comment]
@samikama could someone take a look at this and help debug this issue?

**Comment by samikama** [comment]
Do you still see the issue with recent commits?

**Comment by qinyao-he** [comment]
@samikama I've just tried the most recent master branch. Seems there is error during the conversion.

> 2018-06-26 15:20:51.327926: I tensorflow/core/grappler/devices.cc:51] Number of eligible GPUs (core count >= 8): 1                                                                                 
> 2018-06-26 15:20:51.527677: I tensorflow/contrib/tensorrt/convert/convert_graph.cc:726] MULTIPLE tensorrt candidate conversion: 8                                                                  
> 2018-06-26 15:20:51.527853: I tensorflow/contrib/tensorrt/convert/convert_nodes.cc:2390] Segment @scope '', converted to graph                                                                     
> 2018-06-26 15:20:51.533387: I tensorflow/contrib/tensorrt/convert/convert_nodes.cc:2390] Segment @scope 'dense/', converted to graph                                                               
> 2018-06-26 15:20:51.533767: I tensorflow/contrib/tensorrt/convert/convert_nodes.cc:2390] Segment @scope 'conv2d_transpose/', converted to graph                                                    
> 2018-06-26 15:20:51.534064: I tensorflow/contrib/tensorrt/convert/convert_nodes.cc:2390] Segment @scope 'dropout/dropout/', converted to graph                                                     
> 2018-06-26 15:20:51.534425: I tensorflow/contrib/tensorrt/convert/convert_nodes.cc:2390] Segment @scope 'conv2d_transpose_2/', converted to graph                                                  
> 2018-06-26 15:20:51.534882: I tensorflow/contrib/tensorrt/convert/convert_nodes.cc:2390] Segment @scope '', converted to graph                                                                     
> 2018-06-26 15:20:51.535612: I tensorflow/contrib/tensorrt/convert/convert_nodes.cc:2390] Segment @scope '', converted to graph                                                                     
> 2018-06-26 15:20:51.536245: I tensorflow/contrib/tensorrt/convert/convert_nodes.cc:2390] Segment @scope 'conv2d_transpose_1/', converted to graph                                                  
> 2018-06-26 15:20:51.537682: E tensorflow/contrib/tensorrt/convert/convert_graph.cc:690]  Can't find a GPU device to work with. Please instantiate a session to initialize devices                  
> 2018-06-26 15:20:51.537710: W tensorflow/contrib/tensorrt/convert/convert_graph.cc:812] Can't identify the cuda device. Running on device 0                                                        
> 2018-06-26 15:20:52.558412: E tensorflow/contrib/tensorrt/convert/convert_graph.cc:690]  Can't find a GPU device to work with. Please instantiate a session to initialize devices                  
> 2018-06-26 15:20:52.558439: W tensorflow/contrib/tensorrt/convert/convert_graph.cc:812] Can't identify the cuda device. Running on device 0                                                        
> 2018-06-26 15:20:52.559444: E tensorflow/contrib/tensorrt/log/trt_logger.cc:38] DefaultLogger Parameter check failed at: Network.cpp::addInput::281, condition: isIndexedCHW(dims) && volume(dims) < MAX_TENSOR_SIZE
> 2018-06-26 15:20:52.559459: W tensorflow/contrib/tensorrt/convert/convert_graph.cc:825] Engine creation for segment 1, composed of 3 nodes failed: Invalid argument: Failed to create Input layer tensor InputPH_0 rank=1. Skipping...
> 2018-06-26 15:20:52.559486: E tensorflow/contrib/tensorrt/convert/convert_graph.cc:690]  Can't find a GPU device to work with. Please instantiate a session to initialize devices                  
> 2018-06-26 15:20:52.559491: W tensorflow/contrib/tensorrt/convert/convert_graph.cc:812] Can't identify the cuda device. Running on device 0                                                        
> 2018-06-26 15:20:52.582179: E tensorflow/contrib/tensorrt/convert/convert_graph.cc:690]  Can't find a GPU device to work with. Please instantiate a session to initialize devices                  
> 2018-06-26 15:20:52.582202: W tensorflow/contrib/tensorrt/convert/convert_graph.cc:812] Can't identify the cuda device. Running on device 0                                                        
> 2018-06-26 15:20:52.583201: E tensorflow/contrib/tensorrt/log/trt_logger.cc:38] DefaultLogger Parameter check failed at: Network.cpp::addInput::281, condition: isIndexedCHW(dims) && volume(dims) < MAX_TENSOR_SIZE
> 2018-06-26 15:20:52.583219: W tensorflow/contrib/tensorrt/convert/convert_graph.cc:825] Engine creation for segment 3, composed of 4 nodes failed: Invalid argument: Failed to create Input layer tensor InputPH_0 rank=1. Skipping...
> 2018-06-26 15:20:52.583249: E tensorflow/contrib/tensorrt/convert/convert_graph.cc:690]  Can't find a GPU device to work with. Please instantiate a session to initialize devices                  
> 2018-06-26 15:20:52.583256: W tensorflow/contrib/tensorrt/convert/convert_graph.cc:812] Can't identify the cuda device. Running on device 0                                                        
> 2018-06-26 15:20:52.605704: E tensorflow/contrib/tensorrt/convert/convert_graph.cc:690]  Can't find a GPU device to work with. Please instantiate a session to initialize devices                  
> 2018-06-26 15:20:52.605730: W tensorflow/contrib/tensorrt/convert/convert_graph.cc:812] Can't identify the cuda device. Running on device 0                                                        
> 2018-06-26 15:20:53.000096: E tensorflow/contrib/tensorrt/convert/convert_graph.cc:690]  Can't find a GPU device to work with. Please instantiate a session to initialize devices                  
> 2018-06-26 15:20:53.000124: W tensorflow/contrib/tensorrt/convert/convert_graph.cc:812] Can't identify the cuda device. Running on device 0                                                        
> 2018-06-26 15:20:53.001103: E tensorflow/contrib/tensorrt/log/trt_logger.cc:38] DefaultLogger Parameter check failed at: Network.cpp::addInput::281, condition: isIndexedCHW(dims) && volume(dims) < MAX_TENSOR_SIZE
> 2018-06-26 15:20:53.001119: W tensorflow/contrib/tensorrt/convert/convert_graph.cc:825] Engine creation for segment 6, composed of 3 nodes failed: Invalid argument: Failed to create Input layer tensor InputPH_0 rank=1. Skipping...
> 2018-06-26 15:20:53.001161: E tensorflow/contrib/tensorrt/convert/convert_graph.cc:690]  Can't find a GPU device to work with. Please instantiate a session to initialize devices                  
> 2018-06-26 15:20:53.001166: W tensorflow/contrib/tensorrt/convert/convert_graph.cc:812] Can't identify the cuda device. Running on device 0                                                        
> [1]    20854 segmentation fault (core dumped)  python tensorrt.py

**Comment by tensorflowbutler** [comment]
Nagging Assignee @aaroey: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by aaroey** [comment]
@qinyao-he @xiaocai00 I think the segfault comes from shape inference, when PR #20350 is merged it should go away (I tested on my local machine about that). Now looking at your problem about the conversion.

**Comment by aaroey** [comment]
I ran your script with PR #20350 built with TRT 4.0.1.6, I think all conv layers are now converted:
![0oym5krjj7d](https://user-images.githubusercontent.com/31743510/42596032-18d7f2cc-8509-11e8-941e-aaadd2f3ab48.png)

**Comment by aaroey** [comment]
@qinyao-he @xiaocai00 Would you please wait for PR #20350 to be merged, and build against TRT 4.0.1.6 and try again? Thanks.

**Comment by qinyao-he** [comment]
@aaroey I've just tried your PR branch but is still the same as before. Will this need to be merged to master first?

**Comment by qinyao-he** [comment]
@aaroey I think your result is just the same as mine before. Although there are some trt_op, all conv2d_transpose ops are remains unchanged. So TensorRT still not optimize conv2d_transpose at all.

**Comment by samikama** [comment]
@qinyao-he  are you talking about transpose ops? Transposes are never included in TRT, they are to convert data layouts.

**Comment by qinyao-he** [comment]
@samikama There are no transpose ops. conv2d_transpose is deconv.

**Comment by aaroey** [comment]
@qinyao-he the op type of conv2d_transpose ops is Conv2DBackpropInput, which is currently not supported in the [op list](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/tensorrt/convert/convert_nodes.cc#L2547) of the TF-TRT integration.

**Comment by aaroey** [comment]
@jjsjann123 to check if we would support that op in the future.

**Comment by aaroey** [comment]
@qinyao-he yes that's what I meant. But, you kept mentioning ConvTranspose2D, which nodes have that type? It seems I cannot find any in the graph.

**Comment by qinyao-he** [comment]
@aaroey I think in tensorflow Conv2DTranspose operation (like in 
tf.layers.Conv2DTranspose) is implemented as Conv2DBackpropInput type, but in TensorRT documentation they mentioned as ConvTranspose2D. I think is should be a documentation error, since I think tensorflow doesn't have ConvTranspose2D type. But they should refer to the same thing.

**Comment by aaroey** [comment]
@qinyao-he make sense. I'll leave this to @jjsjann123 to check if/when we can support Conv2DTranspose.

**Comment by ppwwyyxx** [comment]
tensorrt does have "deconvolution" layer support: https://docs.nvidia.com/deeplearning/sdk/tensorrt-api/c_api/classnvinfer1_1_1_i_deconvolution_layer.html , which presumably should map to the "Conv2DBackpropInput" op in tensorflow.

**Comment by jjsjann123** [comment]
TRT does support deconvolution (Conv2DTranspose). We should take a look at this and strat adding support @benbarsdell @pooyadavoodi

**Comment by tensorflowbutler** [comment]
Nagging Assignees @jjsjann123, @aaroey: It has been 20 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @jjsjann123, @aaroey: It has been 35 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @jjsjann123, @aaroey: It has been 50 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @aaroey: It has been 65 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @aaroey: It has been 80 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @smit-hinsu, @aaroey: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by kenryd** [comment]
Have there been any updates on this? It would be very helpful to have deconv supported in tensorflow-tensorrt conversion.

**Comment by pooyadavoodi** [comment]
We are working on this. Stay tuned.
@trevor-m ^^

**Comment by smit-hinsu** [comment]
PR from Trevor to support conv2d_transpose is merged. Thanks, Trevor!
https://github.com/tensorflow/tensorflow/commit/ca2a1cb4cad2dba0d7c4bde662f36de592f5668f

Closing this issue.
